src/analyses/poisonVariables.ml

```diff
@@ -0,0 +1,87 @@
+open Prelude.Ana
+open Analyses
+
+module Spec =
+struct
+  include Analyses.DefaultSpec
+  module VS = SetDomain.ToppedSet(CilType.Varinfo) (struct let topname = "All vars" end)
+
+  let name () = "poisonVariables"
+  module D = VS
+  module C = Lattice.Unit
+
+  let rec check_exp tainted e = match e with
+    (* Recurse over the structure in the expression, returning true if any varinfo appearing in the expression is tainted *)
+    | AddrOf v
+    | StartOf v
+    | Lval v -> check_lval tainted v
+    | BinOp (_,e1,e2,_) -> check_exp tainted e1; check_exp tainted e2
+    | Real e
+    | Imag e
+    | SizeOfE e
+    | AlignOfE e
+    | CastE (_,e)
+    | UnOp (_,e,_) -> check_exp tainted e
+    | SizeOf _ | SizeOfStr _ | Const _  | AlignOf _ | AddrOfLabel _ -> ()
+    | Question (b, t, f, _) -> check_exp tainted b; check_exp tainted t; check_exp tainted f
+  and check_lval ?(ignore_var = false) tainted lval = match lval with
+    | (Var v, offset) ->
+      if not ignore_var && not v.vglob && VS.mem v tainted then M.warn "accessing poisonous variable %a" d_varinfo v;
+      check_offset tainted offset
+    | _ -> () (* TODO: Consider mem *)
+  and check_offset tainted offset = match offset with
+    | NoOffset -> ()
+    | Field (_, o) -> check_offset tainted o
+    | Index (e, o) -> check_exp tainted e; check_offset tainted o
```

Comment:
Seems like this would be a lot easier using a CIL visitor with just a single case for the variables.